src/test/data/script_tests.json

```diff
@@ -2527,6 +2527,344 @@
+    "INVALID_STACK_OPERATION",
+    "TAPSCRIPT Test of OP_CAT flag by calling CAT on an empty stack. This throws an error because OP_CAT flag is set so CAT is executed"
+],
+[
+    [
+        "aa",
+        "bb",
+        "CAT",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT Test of OP_CAT flag by calling CAT on two elements. OP_CAT flag is set so CAT is executed."
+],
+[
+    [
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUAL to stack element 78a11a1260c1101260"
+],
+[
+    [
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT",
+    "OK",
+    "TAPSCRIPT Test of OP_CAT flag, CATs 78a11a1260 and c1101260 together and checks it is EQUAL to stack element 78a11a1260c1101260. No OP_CAT set so CAT should be OP_SUCCESS."
+],
+[
+    [
+        "",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "EVAL_FALSE",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUAL to the empty stack element"
+],
+[
+    [
+        "51",
+        "78a11a1260c1101260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUALVERIFY",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT CATS 78a11a1260 and c1101260 together and checks it is EQUALVERIFY to stack element 78a11a1260c1101260"
+],
+[
+    [
+        "51",
+        "c110126078a11a1260",
+        "78a11a1260",
+        "c1101260",
+        "CAT EQUALVERIFY",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "EQUALVERIFY",
+    "TAPSCRIPT CATs 78a11a1260 and c1101260 together and checks it is EQUALVERIFY to stack element c110126078a11a1260"
+],
+[
+    [
+        "aa",
+        "bb",
+        "CAT 0x4c 0x02 0xaabb EQUAL",
+        "<AUTOGEN:CONTROLBLOCK>",
+        0.00000001
+    ],
+    "",
+    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
+    "P2SH,WITNESS,TAPROOT,OP_CAT",
+    "OK",
+    "TAPSCRIPT CATs aa and bb together and checks EQUAL to aabb"
+],
+[
+    [
+        "eeffeeff",
+        "aa",
+        "bbcc",
+        "CAT CAT DUP DROP 0x4c 0x06 0xeeffeeffaabbcc EQUAL",
```

Comment:
Seems like `0x06` should be `0x07` for the 7 bytes of `0xeeffeeffaabbcc`?
All values `0x00`-`0x07` pass the test successfully, setting it to `0x08` or greater fails the test.
Haven't had the time to dig deeper into https://opcodeexplained.com/opcodes/OP_PUSHDATA1.html (0x4c), but didn't want to sit on this considering it's up for the PR review club tomorrow.

Reply:
Great find!

What is happening is:
1. CAT     -->  ["eeffeeffaa","bbcc"]
2. CAT     -->  ["eeffeeffaabbcc"]
3. DUP    --> ["eeffeeffaabbcc", "eeffeeffaabbcc"]
4. DROP  --> ["eeffeeffaabbcc"]
5. PUSH1 0x06 --> ["eeffeeffaabbcc", "eeffeeffaabb"]
6. 0xcc (OP_SUCCESS204) --> ["eeffeeffaabbcc", "eeffeeffaabb"] 
Exit with success!

You can test this by setting the last byte from 0xcc to 0xff aka but OP_INVALIDOPCODE and it will fail.
`"CAT CAT DUP DROP 0x4c 0x06 0xeeffeeffaaff EQUAL"`
then if you change it to 
`"CAT CAT DUP DROP 0x4c 0x07 0xeeffeeffaaff EQUAL"`
it passes.

I'm going to change this test to `"CAT CAT DUP DROP 0x4c 0x07 0xeeffeeffaaff EQUAL"`

Reply:
Now I see what you mean, before `OP_EQUAL` gets a chance to run, the remaining suffix after the push `0xcc` is executed as an OP-code, and because of `IsOpSuccess()`, it gets treated as `OP_SUCCESS` and script execution stops.

Reply:
0xff is a `OP_NOP` and immediately causes the script the fail with the error "BAD_OPCODE". 

* If the OP_CAT flag is not set then CAT viewed as OP_SUCCESS126, it hits OP_SUCCESS first and returns success.
* If  the OP_CAT flag is set then it makes it to 0xff and fails.

Test it out for yourself:
```cpp
[
    [
        "eeffeeff",
        "aa",
        "bbff",
        "CAT CAT DUP DROP 0x4c 0x06 0xeeffeeffaabbff EQUAL",
        "<AUTOGEN:CONTROLBLOCK>",
        0.00000001
    ],
    "",
    "0x51 0x20 <AUTOGEN:TAPROOTOUTPUT>",
    "P2SH,WITNESS,TAPROOT,OP_CAT",
    "OK",
    "TAPSCRIPT CATs aa and bbcc and eeffeff together and checks EQUAL to eeffeeffaabbcc"
],
```

Reply:
Fixed